larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewTransverseTracksAlgorithm.cc

```diff
@@ -36,19 +55,111 @@ void TwoViewTransverseTracksAlgorithm::CalculateOverlapResult(const Cluster *con
+
+    DiscreteProbabilityVector::InputData<float,float> inputData2;
+    for (const pandora::CaloHit *const pCaloHit: overlapHits2)
+        inputData2.emplace_back(pCaloHit->GetPositionVector().GetX(), pCaloHit->GetInputEnergy());
+
+    const DiscreteProbabilityVector discreteProbabilityVector1(inputData1, xOverlapMax, false);
+    const DiscreteProbabilityVector discreteProbabilityVector2(inputData2, xOverlapMax, false);
+
+    DiscreteProbabilityVector::ResamplingPoints resamplingPointsX;
+    for (unsigned int iSample = 0; iSample < nSamples; ++iSample)
+        resamplingPointsX.emplace_back((xOverlapMin + (xOverlapMax - xOverlapMin) * 
+            static_cast<float>(iSample+1) / static_cast<float>(nSamples+1)));
+
+    const DiscreteProbabilityVector resampledDiscreteProbabilityVector1(discreteProbabilityVector1, resamplingPointsX);
+    const DiscreteProbabilityVector resampledDiscreteProbabilityVector2(discreteProbabilityVector2, resamplingPointsX);
+
+    const float correlation(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(
+        resampledDiscreteProbabilityVector1, resampledDiscreteProbabilityVector2));
+
+    std::mt19937 randomNumberGenerator(static_cast<std::mt19937::result_type>(
+                pCluster2->GetOrderedCaloHitList().size() + pCluster2->GetOrderedCaloHitList().size()));
+
+    const float pvalue(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(
+        resampledDiscreteProbabilityVector1, resampledDiscreteProbabilityVector2, randomNumberGenerator, m_nPermutations));
+
+    const float matchingScore(1.f-pvalue);
 
-    if (xOverlap > std::numeric_limits<float>::epsilon())
-        this->GetMatchingControl().GetOverlapMatrix().SetOverlapResult(pCluster1, pCluster2, xOverlap);
+    const float locallyMatchedFraction(CalculateLocalMatchingFraction(resampledDiscreteProbabilityVector1, 
+        resampledDiscreteProbabilityVector2, randomNumberGenerator));
+
+    overlapResult = TwoViewTransverseOverlapResult(matchingScore, resampledDiscreteProbabilityVector1.GetSize(), 
+        correlation, locallyMatchedFraction, twoViewXOverlap);
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float TwoViewTransverseTracksAlgorithm::CalculateLocalMatchingFraction(const DiscreteProbabilityVector &discreteProbabilityVector1,
+    const DiscreteProbabilityVector &discreteProbabilityVector2, std::mt19937 &randomNumberGenerator)
+{
+    if (discreteProbabilityVector1.GetSize() != discreteProbabilityVector2.GetSize() || 
+            0 == discreteProbabilityVector1.GetSize()*discreteProbabilityVector2.GetSize())
+        throw STATUS_CODE_INVALID_PARAMETER;
+
+    if (m_minSamples > discreteProbabilityVector1.GetSize())
+        throw STATUS_CODE_INVALID_PARAMETER;
+
+    std::vector<float> localValues1, localValues2;
```

Comment:
Could use FloatVector, for consistency (only argument, really).

Reply:
Yepper. Done.